File: 백준/Gold/18231. 파괴된 도시/파괴된 도시.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        StringBuilder sb = new StringBuilder();

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int[] visited = new int[N + 1];
        ArrayList<Integer> broken = new ArrayList<>();
        ArrayList<Integer> result = new ArrayList<>();

        List<List<Integer>> graph = new ArrayList<>(N + 1);
        for (int i = 0; i <= N; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            graph.get(a).add(b);
            graph.get(b).add(a);
        }

        int K = Integer.parseInt(br.readLine());

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < K; i++) {
            int num = Integer.parseInt(st.nextToken());
            broken.add(num);
            visited[num] = 1;
        }

        HashSet<Integer> possible = new HashSet<>();

        for (int i: broken) {
            boolean check = false;

            for (int j : graph.get(i)) {
                if (visited[j] == 0) {
                    check = true;
                    break;
                }
            }

            if (!check) {
                possible.add(i);
                result.add(i);
                possible.addAll(graph.get(i));
            }
        }

        if (possible.size() != K) {
            System.out.println(-1);
        } else {
            System.out.println(result.size());
            for (int i = 0; i < result.size(); i++) {
                sb.append(result.get(i)).append(" ");
            }
            System.out.println(sb);
        }
    }
}

File: 백준/Gold/18231. 파괴된 도시/파괴된 도시.py
# 파괴된 도시
import sys
input = sys.stdin.readline

N, M = map(int,input().split())
graph = [[] for _ in range(N+1)]
visited = [0] * (N+1)
result = []

for i in range(M):
    a, b = map(int,input().split())
    graph[a].append(b)
    graph[b].append(a)

K = int(input())
broken = list(map(int,input().split()))
broken.sort()

# 파괴된 도시를 따로 체크
for i in broken:
    visited[i] = 1

# 폭탄 맞은 도시를 저장
possible = set()

for i in broken:
    check = False
    
    for j in graph[i]:
        # 주변도시가 파괴되지 않앗다면 폭탄 맞은 도시가 아니다
        if visited[j] == 0:
            check = True
            break

    if not check:
        # 폭탄 맞은 도시일 경우 주변 도시도 파괴되니 저장
        possible.add(i)
        result.append(i)
        for j in graph[i]:
            possible.add(j)

# 주어진 정보와 계산한 정보가 다르면 -1
if len(possible) != K:
    print(-1)
else:
    print(len(result))
    print(*result)
